First release

nschan opened the pull request:
This is a release PR to merge dev into master. I have accidentally already created main from dev, if main is preferred over master I can remove main and open a PR.

This pipeline was originally initiated by @mahesh-panchal I think. This is however a complete rewrite from scratch of the pipeline.

## PR checklist

- [ x] This comment contains a description of changes (with reason).
- [x ] If you've fixed a bug or added code that should be tested, add tests!
- [x ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/genomeassembler/tree/master/.github/CONTRIBUTING.md)
- [x ] If necessary, also make a PR on the nf-core/genomeassembler _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ x] Make sure your code lints (`nf-core pipelines lint`).
- [x ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ x] Usage Documentation in `docs/usage.md` is updated.
- [ x] Output Documentation in `docs/output.md` is updated.
- [x ] `CHANGELOG.md` is updated.
- [ x] `README.md` is updated (including new tool citations and authors/contributors).

d4straub commented on README.md:
```diff
@@ -1,76 +1,126 @@
+| Yes | Yes   | hifiasm --ul                                       | `hifiont_hifiasm`     |
+| Yes | Yes   | Scaffolding of ONT assemblies onto HiFi assemblies | `hifiont_flyehifiasm` |
+
+### Pipeline specific parameters
```
This whole section isnt supposed to be here in the `README.md` (I would vote for deleting it) but rather in `nextflow_schema.json` (amend if neccessary, but there are some entries already, didnt check for completeness) so that it will be presented in https://nf-co.re/genomeassembler/parameters/.

d4straub commented on README.md:
If possible, it would be great to have some sort of visual representation of the pipeline, here are some guidelines and ideas: https://nf-co.re/docs/guidelines/graphic_design/workflow_diagrams

d4straub commented on docs/output.md:
This looks a little different to the usual documentation, for example https://nf-co.re/mag/output/, https://nf-co.re/rnaseq/output/
For example, I would have hoped in section `Pipeline overview` for headlines such as `QC`, `Assembly`, `Assembly post-processing`, and `Assembly QC` or such. Just a thought

nschan replied:
I have reformatted this file, I hope it is closer to your expectations now?

d4straub commented on docs.md:
That seems redundant with other documentation is seems not part of the typical nf-core file structure?

d4straub commented on parameters.md:
Empty file can be removed I think

d4straub commented on schema.md:
That seems redundant with other documentation is seems not part of the typical nf-core file structure?
According to standards, that should be documented in `nextflow_schema.json`.

d4straub reviewed (commented):
Hi, I only had a look at documentation for now and I have some comments and suggestions.

LouisLeNezet commented on modules/local/liftoff/main.nf:
Shouldn't this one be deleted to only use the one from nf-core ?

nschan replied:
One of them should be deleted, yes. I am actually only using my local module, but I think I can switch to the nf-core module if that is strongly preferred. 
To provide some explanation for some of my local modules: For many of the tools that use references, I am facing an issue where nf-core modules (as I understand) assume a global reference that comes from a param, but in this pipeline the reference comes with the sample, so different samples can use different references. This is of course not a total blocker, but it requires me to do a lot of channel splitting and joining to make sure that sample and reference stay paired correctly when using some of the nf-core modules, and sometimes it seems more pragmatic to have a local module that expects the channels I have.

LouisLeNezet replied:
Okay, make sense.
I had some trouble like that a time ago, and I used the `nf-core modules patch` command to fix it.
Doing so you have the best of both world 😉

LouisLeNezet commented on parameters.md:
What is this file ?

LouisLeNezet reviewed (changes requested):
Hi,
You've done quite the work. I didn't went through the whole detail yet of the pipeline however here is a few comments and some suggestion.

- It seems that you have been directly commiting to the dev branch for some time now. I wouldn't recommend this practice has there is no trace in the changelog and no review by anyone else. What you should do is for each modification / new addition, create a new branch and do a PR to the dev branch as you did for the PR #62.
- Whenever possible you should use the module from the nf-core repository or add them to it if absent. This way your work would benefit everyone (e.g. hifiasm, jellyfish, ...). My point of view is that in the local module directory should be only pipeline really specific process (graph or data transformation) that no one else would use.
- Some of the process you've wrote differ only by a few argument. I would recommend you to write the process as versatile as possible and change its use through a config file.
- It would be nice to have nf-test for your different subworkflow to ensure that everything work as expected.

d4straub commented on CHANGELOG.md:
```diff
@@ -3,7 +3,7 @@
-## v1.0dev - [date]
+## v1.0.0dev - [date]
 
 Initial release of nf-core/genomeassembler, created with the [nf-core](https://nf-co.re/) template.
```
I disagree, that isnt relevant for the first release, imho, but only for subsequent releases to make changes more obvious.

d4straub commented on CHANGELOG.md:
```diff
@@ -3,7 +3,7 @@
 
-## v1.0dev - [date]
+## v1.0.0dev - [date]
 
```
Also here, for a first release maybe not relevant.

nschan replied:
There were no PRs or versions, I wrote the whole thing and this is the first release version.

nschan commented on assets/report/functions/read_nanoq.R:
```diff
@@ -0,0 +1,29 @@
+library(magrittr)
```
Historic reasons, this function came from some other project of mine, I have modified it.

nschan commented:
> Hi, You've done quite the work. I didn't went through the whole detail yet of the pipeline however here is a few comments and some suggestion.

Thanks a lot!

> * It seems that you have been directly commiting to the dev branch for some time now. I wouldn't recommend this practice has there is no trace in the changelog and no review by anyone else. What you should do is for each modification / new addition, create a new branch and do a PR to the dev branch as you did for the PR [Assembler #62](https://github.com/nf-core/genomeassembler/pull/62).

I developed in the assembler branch, which itself was a migration of my initial non nf-core pipeline into the rather empty genomeassembler repo / nf-core. I merged assembler into dev with #62. I did not open PRs against my assembler branch for my commits (that seemed a bit unnecessary), if that is what you are referring to.

> * Whenever possible you should use the module from the nf-core repository or add them to it if absent. This way your work would benefit everyone (e.g. hifiasm, jellyfish, ...). My point of view is that in the local module directory should be only pipeline really specific process (graph or data transformation) that no one else would use.

This is something that I tried to clarify in #62, I guess opinions differ regarding non nf-core modules; I will take another look which modules can / should be nf-core. I tend to write modules that I think are OK for me but shouldn't be inflicted on the larger community.

> * Some of the process you've wrote differ only by a few argument. I would recommend you to write the process as versatile as possible and change its use through a config file.

Good point, I wrote some of those modules before I understood how modules.conf works. It probably makes sense for me to review my local modules, also with regards to your previous comment.

> * It would be nice to have nf-test for your different subworkflow to ensure that everything work as expected.

I will take a look at this.

nschan commented on README.md:
```diff
@@ -1,76 +1,126 @@
-## Documentation
-
-The nf-core/genomeassembler pipeline comes with documentation about the pipeline [usage](https://nf-co.re/genomeassembler/usage), [parameters](https://nf-co.re/genomeassembler/parameters) and [output](https://nf-co.re/genomeassembler/output).
+**nf-core/genomeassembler** is a bioinformatics pipeline that carries out genome assembly, polishing and scaffolding from long reads (ONT or pacbio). Assembly can be done via `flye` or `hifiasm`, polishing can be carried out with `medaka` (ONT), or `pilon` (requires short-reads), and scaffolding can be done using `LINKS`, `Longstitch`, or `RagTag` (if a reference is available). Quality control in ...
```
Thanks.

nschan commented on assets/report/report.qmd:
Due to the somewhat convoluted approach quarto takes to conditional content, this file is probably difficult to review in any case. I will start by adding comments to this so it is a bit less hard to follow.

LouisLeNezet commented on modules/local/hifiasm/main.nf:
This is typically a good reason to update the nf-core modules. What I would recommend would be to set a new optional input file and add it to the main script.
I don't know if the `--ul` option is exclusive with other options but you could do something like:

```
input:
    tuple val(meta) , path(reads), path(reads_ont)
    tuple val(meta1), path(paternal_kmer_dump), path(maternal_kmer_dump)
    tuple val(meta2), path(hic_read1)         , path(hic_read2)
...
def cmd_reads_ont = reads_ont ? "--ul ${reads_ont}" : ""
...
 hifiasm \\
            $args \\
            -o ${prefix}.asm \\
            -t $task.cpus \\
            $cmd_reads_ont \\
            $reads \\
            2> >( tee ${prefix}.stderr.log >&2 )
```

muffato replied:
There may also be the option of passing `--ul` through `ext.args`

nschan replied:
>There may also be the option of passing --ul through ext.args
`--ul` yes, but I dont think I can pass extra input through `ext.args`?

LouisLeNezet replied:
As the `--ul` input is a file it should be added to the input channel of the nf-core modules, as said in the [guidelines](https://nf-co.re/docs/guidelines/components/modules).

> All mandatory and optional input files MUST be included in input channel definitions.

Doing this update would benefit everyone after that would like to use this module with ONT reads.

nschan replied:
I will update the nf-core module next week based on your suggestions.

LouisLeNezet replied:
From my point of view this is a good thing to be done.
Normally each module should declare all the different files argument. The ONT reads was missing and need to be added.
If you look at the bcftools modules, they have plenty of argument, and depending of your use case you will effectively use empty list. Until there is named arguments in nextflow that will be kept as such I think.

nschan replied:
The [hifiasm module was updated](https://github.com/nf-core/modules/pull/7336).

LouisLeNezet commented on README.md:
```diff
@@ -1,76 +1,76 @@
+Each row represents one genome to be assembled. `sample` should contain the name of the sample, `ontreads` should contain a path to ONT reads (fastq.gz), `hifireads` a path to HiFi reads (fastq.gz), `ref_fasta` and `ref_gff` contain reference genome fasta and annotations. `shortread_F` and `shortread_R` contain paths to short-read data, `paired` indicates if short-reads are paired. Columns can be ...
 
-2. Install any of [`Docker`](https://docs.docker.com/engine/installation/), [`Singularity`](https://www.sylabs.io/guides/3.0/user-guide/) (you can follow [this tutorial](https://singularity-tutorial.github.io/01-installation/)), [`Podman`](https://podman.io/), [`Shifter`](https://nersc.gitlab.io/development/shifter/how-to-use/) or [`Charliecloud`](https://hpc.github.io/charliecloud/) for full pip ...
+-->
```
```suggestion

```

LouisLeNezet commented on conf/modules.config:
```diff
@@ -17,25 +17,438 @@ process {
+    withName: '.*MAP_TO_ASSEMBLY.*' {
+        ext.prefix = { "${meta.id}_${task.process.tokenize(':')[2].toLowerCase()}" }
+        publishDir = [
+            path: { "${params.outdir}/QC/${task.process.tokenize(':')[2].toLowerCase()}${task.process.tokenize(':')[2].toLowerCase() == 'assemble' ? '' : '/' + task.process.tokenize(':')[-3].tokenize('_')[1].toLowerCase()}/quast/${meta.id}/map_to_assembly/${task.process.tokenize(':')[-1].tokenize('_')[0].toLowerCase()}" },
```
This is quite the complex path.
Did you test it to check if it was correctly published ?

nschan replied:
It works ok, but I have decided to make this more explicit, since it is a confusing path and will probably still be confusing in the future..

LouisLeNezet commented on docs/output.md:
```diff
@@ -2,56 +2,274 @@
-![MultiQC - FastQC adapter content plot](images/mqc_fastqc_adapter.png)
+- `assemble/`
+  - `flye/`: output from flye.
+    - `<SampleName>/`
```
Didn't you change the organisation by `SampleName/tool` ?

nschan replied:
Yes, but this PR was not merged when you reviewed :)

LouisLeNezet commented on docs/output.md:
```diff
@@ -2,56 +2,274 @@
 ## Introduction
 
-This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.
+This document describes the output produced by the pipeline..
```
This is out of the scope of this PR, but it could be interesting to have your plot generation added to MultiQC 😉

nschan replied:
That is a good idea, I have not used MultiQC and am not familiar with how to extend it, but I will take a look if I have some time available

LouisLeNezet commented on docs/usage.md:
```diff
@@ -6,106 +6,148 @@
+
+Pipeline settings can be provided in a `yaml` or `json` file via `-params-file <file>`.
+
+:::warning
```
You might prefer here to use

> [!WARNING]  
> My note

LouisLeNezet commented on docs/usage.md:
```diff
@@ -6,106 +6,148 @@
+To further assist in reproducbility, you can use share and re-use [parameter files](#running-the-pipeline) to repeat pipeline runs with the same settings without having to write out a command with every single parameter.
 
-This version number will be logged in reports when you run the pipeline, so that you'll know what you used when you look back in the future.
+:::tip
```
Same here, prefer the [!WARNING/TIP/NOTE/IMPORTANT/CAUTION] for a sure rendering.

LouisLeNezet commented on docs/usage.md:
```diff
@@ -6,106 +6,148 @@
 ## Core Nextflow arguments
 
-> **NB:** These options are part of Nextflow and use a _single_ hyphen (pipeline parameters use a double-hyphen).
+:::note
```
Same as above

LouisLeNezet commented on docs/usage.md:
```diff
@@ -6,106 +6,148 @@
+Several generic profiles are bundled with the pipeline which instruct the pipeline to use software packaged using different methods (Docker, Singularity, Podman, Shifter, Charliecloud, Apptainer, Conda) - see below.
 
-> We highly recommend the use of Docker or Singularity containers for full pipeline reproducibility, however when this is not possible, Conda is also supported.
+:::info
```
Same as above

LouisLeNezet commented on modules/local/gfa2fa/enivronment.yml:
Shouldn't it be `environment.yml` ?

nschan replied:
Yes...

LouisLeNezet commented on modules/nf-core/liftoff/liftoff.diff:
```diff
@@ -0,0 +1,53 @@
+     def arg_db  = ref_db            ?   "-db $ref_db"           : ''
+     prefix      = task.ext.prefix   ?:  "${meta.id}"
+     """
++    if [[ ${target_fa} == *.gz ]]; then
```
This could be a nice addition to the nf-core module.

LouisLeNezet commented on subworkflows/local/assemble/main.nf:
```diff
@@ -0,0 +1,222 @@
+    // Define inputs for flye
+    if (params.assembler == "flye") {
+      if (params.hifi) {
+        if (!hifi_only) {
```
This could be also checked at the `PIPELINE_INITIALISATION` step. But as there is no process ran before hand there might not be any need. The aim is to throw an error the quickest as possible if the user does not specify the correct set of argument.

LouisLeNezet commented on subworkflows/local/assemble/main.nf:
```diff
@@ -0,0 +1,222 @@
+      // When using either hifiasm_ont or flye_on_hifiasm, both reads are available, which should be used for qc?
+      if (params.hifi && params.ont) {
+        if (!params.qc_reads) {
+          error("Please specify which reads should be used for qc: 'ONT' or 'HIFI'")
```
This one could be throw earlier in the pipeline.

LouisLeNezet commented on subworkflows/local/qc/merqury/main.nf:
```diff
@@ -0,0 +1,30 @@
+    assembly.map {meta, _assembly -> [meta.id, []]}.set { spectra_cn_hist }
+    assembly.map {meta, _assembly -> [meta.id, []]}.set { assembly_qv }
+    if (params.merqury) {
+    meryl_out
```
There is a missing indentation here

LouisLeNezet reviewed (changes requested):
Thanks for addressing my past comments.
Here is some new ones before approving the first release.

LouisLeNezet commented:
Hi @nschan,
If there is a issue with the pipeline TEMPLATE syncing I would suggest you to ask on slack.
I had some trouble with our own pipeline before and people on slack was really helpfull.
From my point of view it would be better to fix the TEMPLATE issue before merging, but that is out of my competence scope.
I'll try to review in the coming days.

nschan commented:
> From my point of view it would be better to fix the TEMPLATE issue before merging, but that is out of my competence scope.
Yes, I agree. I wanted to discuss if, after review is done, I can close this PR, fix the template branch and then open a new PR & merge that (without going through another release-review), instead of attempting to fix the TEMPLATE branch in here, which I think might actually not work..?

nschan commented on modules/local/align/main.nf:
```diff
@@ -0,0 +1,75 @@
+    """
+}
+
+process ALIGN_SHORT_TO_BAM {
```
I have switched to the nf-core module for minimap2_align.

mirpedrol commented on .nf-core.yml:
```diff
@@ -1 +1,16 @@
+  files_exist:
+    - conf/igenomes.config
+    - conf/igenomes_ignored.config
+    - ro-crate-metadata.json
```
the ro-crate file linting shouldn't be skipped, you can generate a ro-crate file with `nf-core pipelines rocrate`

nschan replied:
Ok, done, see #89.

mirpedrol commented on CITATIONS.md:
```diff
@@ -10,10 +10,101 @@
 
-- [MultiQC](https://pubmed.ncbi.nlm.nih.gov/27312411/)
-  > Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016 Oct 1;32(19):3047-8. doi: 10.1093/bioinformatics/btw354. Epub 2016 Jun 16. PubMed PMID: 27312411; PubMed Central PMCID: PMC5039924.
+- [lima](https://github.com/pacificbiosciences/barcoding/)
```
isn't there a citation for lima?

nschan replied:
I couldnt find one on the pacbio website(s) (the one provided and https://lima.how/)

mirpedrol commented on CITATIONS.md:
```diff
@@ -10,10 +10,101 @@
+
+  > Steinig and Coin (2022). Nanoq: ultra-fast quality control for nanopore reads. Journal of Open Source Software, 7(69), 2991, https://doi.org/10.21105/joss.02991
+
+- [porechop](https://github.com/rrwick/Porechop)
```
and the same for porechop

mirpedrol commented on CITATIONS.md:
```diff
@@ -10,10 +10,101 @@
+
+  > Bruce J. Walker, Thomas Abeel, Terrance Shea, Margaret Priest, Amr Abouelliel, Sharadha Sakthikumar, Christina A. Cuomo, Qiandong Zeng, Jennifer Wortman, Sarah K. Young, Ashlee M. Earl (2014) Pilon: An Integrated Tool for Comprehensive Microbial Variant Detection and Genome Assembly Improvement. PLoS ONE 9(11): e112963. doi:10.1371/journal.pone.0112963
+
+- [medaka](https://github.com/nanoporetech/medaka)
```
citation?

nschan replied:
I dont think ont formally published medaka.

mirpedrol commented on README.md:
```diff
@@ -1,76 +1,77 @@
 ## Credits
 
-nf-core/genomeassembler was originally written by @d4straub.
+nf-core/genomeassembler was originally written by Niklas Schandry (@nschan).
```
I'm not sure about the history of this pipeline, but wouldn't it be better to say something like "originally written by @d4straub and re-written/modified by @nschan"? 
And you can also add your affiliation.
This is just a suggestion but of course I leave it for you to decide the best way to express contributions 🙂

nschan replied:
I wrote this pipeline from scratch. There was a `genomeassembler` dev, which was pretty empty and I replaced everything in there when I decided to contribute this pipeline to nf-core. Unless @d4straub disagrees I think it is correct as it is currently written. I will add my affiliation 👍 .

mirpedrol commented on assets/adaptivecard.json:
just for me to understand, what is this JSON doing and where is it used?

nschan replied:
I have no idea what it does, I assumed it was part of nf-core tools setup, if it is not needed I can remove it.

muffato replied:
This file comes from the pipeline template. It's used to report completion of the pipeline in Teams

mirpedrol commented on assets/multiqc_config.yml:
you are not using multiQC, right? Then you can remove this file

nschan replied:
This is in #91

mirpedrol commented on assets/report/_quarto.yml:
where are these files used?

mirpedrol replied:
thanks for the explanation 👍

mirpedrol commented on assets/schema_input.json:
```diff
@@ -10,15 +10,40 @@
+                "format": "file-path",
                 "pattern": "^\\S+\\.f(ast)?q\\.gz$",
-                "errorMessage": "FastQ file for reads 1 must be provided, cannot contain spaces and must have extension '.fq.gz' or '.fastq.gz'"
+                "errorMessage": "ONT reads must be provided, cannot contain spaces and must have extension '.fq.gz' or '.fastq.gz'"
```
From the `must be provided` sentence I understand that these parameters are required, but they are not provided in the `required` field, so I would remove this bit from all the error messages.
```suggestion
                "errorMessage": "ONT reads cannot contain spaces and must have extension '.fq.gz' or '.fastq.gz'"
```

nschan replied:
Good point, I have updated this file, see #93

mirpedrol commented on bin/genomescope.R:
```diff
@@ -0,0 +1,752 @@
+## This is the automated script for computing genome characteristics
+## from a kmer histogram file, k-mer size, and readlength
+
+## https://github.com/schatzlab/genomescope/
```
the license should be specified in the header, see https://github.com/nf-core/crisprseq/blob/master/bin/cigar_parser.R#L7 for an example. The other information that you already have looks good 👍

mirpedrol commented on conf/igenomes_ignored.config:
since you are not using ignomes at all, you can also remove this file and ignore it in the linting. The same as you did for `igenomes.config`

mirpedrol reviewed (commented):
Posting a first round of comments, I am half way reviewing, will continue later today!

A general comment. Since I think you are not using FastQC nor MultiQC, you can remove this modules `nf-core modules remove`. And you can add them to the features skipped in the `.nf-core.yml` file to avoid the template updates adding them in the future, see [customisation](https://nf-co.re/docs/nf-core-tools/pipelines/create#customization) in the docs.

mirpedrol commented on configs/hifi_flye.config:
```diff
@@ -0,0 +1,6 @@
+params {
```
a comment describing what this config is for would be good, and the same for your other cunstom configs

nschan replied:
That is a good idea, done in #96

mirpedrol commented on docs/usage.md:
```diff
@@ -116,11 +154,25 @@ If `-profile` is not specified, the pipeline will run locally and expect all sof
+
+To ease configuration, there are a couple of pre-defined profiles for various combinations of read sources and assemblers (named readtype_assembler)
+
+| ONT | HiFI  | Assembly strategy                                  | Profile name          |
+| --- | ----- | -------------------------------------------------- | --------------------- |
+| Yes | No    | flye                                               | `ont_flye`            |
+| Yes | No    | hifiasm                                            | `ont_hifiasm`         |
+| No  | Yes   | flye                                               | `hifi_flye`           |
+| No  | Yes   | hifiasm                                            | `hifi_hifiasm`        |
+| Yes | Yes   | hifiasm --ul                                       | `hifiont_hifiasm`     |
+| Yes | Yes   | Scaffolding of ONT assemblies onto HiFi assemblies | `hifiont_flyehifiasm` |
```
this is duplicated above, in the first section of this file

nschan replied:
Bundled this to the config comments in #96

mirpedrol commented on modules/local/genomescope/main.nf:
```diff
@@ -0,0 +1,39 @@
+        | grep 'Haploid Length' \\
+        | sed 's@ bp@@g' \\
+        | sed 's@,@@g' \\
+        | awk '{printf "%i", (\$4+\$5)/2 }')
```
it would be nice if this module outputs the version like nf-core modules

mirpedrol commented on modules/local/jellyfish/main.nf:
```diff
@@ -0,0 +1,115 @@
+            -C \\
+            -t ${task.cpus} ${fasta.baseName}.fasta
+
+        mv mer_counts.jf ${prefix}_mer_counts.jf
```
same, you could output the version of the tool

mirpedrol commented on modules/local/links/main.nf:
```diff
@@ -0,0 +1,34 @@
+    """
+    echo "${reads}" > readfile.fof
+    LINKS -f ${assembly} -s readfile.fof -j 3 -b ${prefix}_links -t 40,200 -d 500,2000,5000
+    sed -i 's/\\(scaffold[0-9]*\\).*/\\1/' ${prefix}_links.scaffolds.fa
```
output version

mirpedrol commented on modules/local/longstitch/main.nf:
```diff
@@ -0,0 +1,52 @@
+        cp ${reads} reads.fq.gz
+    fi
+
+    longstitch tigmint-ntLink-arks draft=assembly reads=reads t=${task.cpus} G=135e6 out_prefix=${prefix}
```
output version

mirpedrol commented on modules/local/medaka/main.nf:
there is a medaka module in nf-core, is there a reason why you are using a local module?

nschan replied:
The nf-core module is version medaka:1.4.4; I am using 2.0.1 in `medaka_consensus`; this main.nf is not imported, I should probably delete it.

mirpedrol replied:
you can also update the version in the nf-core/module. It's ok for this release, but I encourage you to do it for the next one

mirpedrol commented on modules/local/nanoq/main.nf:
```diff
@@ -0,0 +1,37 @@
+    script:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    nanoq -i ${reads} -j -r ${prefix}_report.json -s -H -vvv > ${prefix}_stats.json
```
output the version, you already too in the stub but not in the main script, and it is not an output channel

mirpedrol commented on modules/local/ragtag/main.nf:
```diff
@@ -0,0 +1,48 @@
+        mv ${assembly} ${prefix}.fa
+    fi
+
+    ragtag.py scaffold ${reference} ${prefix}.fa \\
```
version

mirpedrol commented on modules/local/report/main.nf:
```diff
@@ -0,0 +1,75 @@
+    }
+    """
+    export HOME="\$PWD"
+    quarto render report.qmd \\
```
version

mirpedrol commented on nextflow.config:
```diff
@@ -9,126 +9,307 @@
     // References
+    /*
     genome                     = null
-    igenomes_base              = 's3://ngi-igenomes/igenomes'
-    igenomes_ignore            = false
-
+    igenomes_base              = 's3://ngi-igenomes/igenomes/'
+    */
+    //igenomes_ignore            = false
     // MultiQC options
     multiqc_config             = null
     multiqc_title              = null
+    multiqc_logo               = null
     max_multiqc_email_size     = '25.MB'
+    multiqc_methods_description = null
```
this can be deleted if you don't use igenomes nor multiqc

mirpedrol commented on nextflow.config:
```diff
@@ -9,126 +9,307 @@
 
+    // Pipeline params
+    input                      = '' // input file
+    outdir                     = './results' // outdir
```
we usually leave the outdir unset, and define it as a required parameter in the nextflow parameters schema

nschan replied:
Ok, this is #102

mirpedrol commented on nextflow_schema.json:
```diff
@@ -213,57 +142,318 @@
+                    "help_text": "Incoming hook URL for messaging service. Currently, MS Teams and Slack are supported.",
+                    "hidden": true
+                },
                 "multiqc_config": {
```
multiqc parameters can be removed

nschan replied:
See #99

mirpedrol commented on subworkflows/local/assemble/main.nf:
```diff
@@ -0,0 +1,218 @@
+            }
+            // Run flye
+            FLYE(flye_inputs, params.flye_mode)
+            FLYE.out.fasta.set { ch_assembly }
```
you should take the versions of all modules and add them to a versions channel. See an example from [rnaseq](https://github.com/nf-core/rnaseq/blob/master/workflows/rnaseq/main.nf#L211) or any other nf-core pipeline

mirpedrol replied:
this applies to all other modules, I won't add a comment to all of them 😄

nschan replied:
I will go through the local modules and have them emit versions and collect all versions into the versions channel, probably next week.

nschan replied:
This is in #101

mirpedrol commented on subworkflows/local/polishing/medaka/run_medaka/main.nf:
I don't see the need to create subworkflows if they only run one single module. Unless you are reusing the code somewhere in the code that I haven't seen. That said, this is not a blocker for this PR, but just a recommendation.

nschan replied:
Fair point. I created these subworkflows mainly to deal with the joining of channels inside the subworkflow, to be able to easily copy the polish_tool workflows. 
Looking back I am not so sure if that was really necessary, in this pipeline I am somewhat frequently encountering the issue that many nf-core modules like `reference_genome` as a separate input, but since I would like to be able to switch reference per sample i need to join reference and assembly quite often, which i hide in the subworkflow :).

mirpedrol replied:
Ok that makes sense. 
Just in case you are doing this, if you have one channel with reads and reference for example, and need to split it into two different channels, one for reads and one for reference, to use the module. The safest option is to use `multiMap`, this will ensure that the order of the channels is the correct one, otherwise, as nextflow channels are FIFO, you could end up using the wrong reference.

mirpedrol reviewed (commented):
Last round of comments 😄

nschan commented:
Sorry for proposing a change after review, but I noticed that the current output structure is kind of confusing. 
It currently goes something like `results/subworkflow/tool/sample`, but I think that `results/sample/subworkflow/tool` would be more intuitive, since I assume most users actually want to look at all outputs from one sample, instead of all outputs per tool? What do you think @mirpedrol @LouisLeNezet ? If it is ok for you I update modules.conf and output.md?

mirpedrol commented:
Sounds good, you can change the output directories 🙂 some pipelines also group them by a more meaningful name, instead of using the subworkflow+tool name, but that's up to you to decide what would be best for your users.

nschan commented:
The module output updates in #100

nschan commented:
Hi @LouisLeNezet, github tells me that you requested changes, but I am a bit lost in the review and I cannot tell what is that you requested that I missed. Could you let me know again what I should change, or if this has been resolved meanwhile?
I think I addressed all points raised by @mirpedrol (open PRs #100, #101 and #102 ready for review), and would like to finalise the review process.

nschan commented on .github/workflows/ci.yml:
```diff
@@ -7,46 +7,48 @@ on:
+          - "ont_hifiasm"
+          - "hifiont_hifiasm"
+          - "hifiont_flyehifiasm"
+        profile: ["conda", "docker"]
```
I recall adding singularity, and those tests always failed, I think it was problems related to the runners?

nschan replied:
I added them back in experimentally in #104 (they fail).

nschan replied:
I merged #105, so they now also fail here :/

mirpedrol replied:
You need to install singularity within the `ci.yml` file, this was probably removed when you removed the `singularity` option.
Please see the [file](https://github.com/nf-core/tools/blob/main/nf_core/pipeline-template/.github/workflows/ci.yml) from the template for reference. Especially the lines from 57 to 65

nschan replied:
Thanks a lot @mirpedrol; what about the conda installation steps, should I also add those (both?)?

mirpedrol replied:
yes, this file should look exactly like the one in the template, except from your custom test names that you added

nschan commented on modules/local/jellyfish/main.nf:
```diff
@@ -0,0 +1,115 @@
+process COUNT {
```
I am happy if someone takes on adding this to nf-core during the hackathon (not sure if/how I can propose stuff for the hackathon). I myself will not be able to participate since I have teaching obligations that overlap.
I will separate those modules into their own folders.

nschan commented on .github/workflows/release-announcements.yml:
```diff
@@ -0,0 +1,75 @@
+          python-version: "3.10"
+      - name: Install dependencies
+        run: pip install tweepy==4.14.0
+      - name: Send tweet
```
It should be 3.2.0 per #71

nschan commented on subworkflows/local/mapping/map_to_assembly/main.nf:
```diff
@@ -0,0 +1,35 @@
+    emit:
+    aln_to_assembly_bam
+    aln_to_assembly_bai
+    aln_to_assembly_bam_bai
```
`aln_to_assembly_bam` is actually sufficient, yes. I can remove the other two. I would prefer to do this once #101 and #100 are merged into dev, since both modify some modules and having a third branch that modifies module code seems like I will have to spend some time resolving conflicts ;)

nschan replied:
This is in #103, please lmk if thats ok @LouisLeNezet

mirpedrol commented on subworkflows/local/utils_nfcore_genomeassembler_pipeline/main.nf:
```diff
@@ -0,0 +1,267 @@
+    // Can use ternary operators to dynamically construct based conditions, e.g. params["run_xyz"] ? "<li>Author (2023) Pub name, Journal, DOI</li>" : "",
+    // Uncomment function in methodsDescriptionText to render in MultiQC report
+    def reference_text = [
+        "<li>Andrews S, (2010) FastQC, URL: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/).</li>",
```
This is not used in the pipeline, see lines 256-258 commented out, so it's not needed to add the references. But it's a nice suggestion to use it in the future 🙂

nschan commented on workflows/genomeassembler.nf:
```diff
@@ -1,123 +1,224 @@
-    CUSTOM_DUMPSOFTWAREVERSIONS (
-        ch_versions.unique().collectFile(name: 'collated_versions.yml')
-    )
+    REPORT(report_files, report_functions, nanoq_files, genomescope_files, quast_files, busco_files, merqury_files)
```
I have added this to #101

nschan commented:
Thanks for reviewing. To make sure (because I am not seeing anything unresolved): are there any open points from your reviews that that I have missed @LouisLeNezet @mirpedrol ?

nschan commented:
Download pipelines still fails as a GHA. When looking at the output I noticed that the stub run fails, due to a broken stub in gfa2fa (local). This stub is fixed in #106

mirpedrol reviewed (approved):
Thanks for addressing all the comments 🙂  I will approve the PR, but please make sure that all tests pass before merging

mirpedrol commented:
It's OK to ignore the branch protection test failing, it is fixed in the PR (missing `=` in the expression), and it will pass in the following PRs once the fix is merged to master 👍

nschan commented:
> Thanks for addressing all the comments 🙂 I will approve the PR, but please make sure that all tests pass before merging

Thanks. I do not know what to do about pipelines download to be honest. @mashehu tried running the test in a gitpod and it worked fine, so I assume it is related to the GHA somehow? From what I picked up on slack it is not only here that this test causes problems and it seems to also be flaky for some other pipelines.
There was another broken stub, this is in #107

mirpedrol commented:
The download test is currently failing because it is missing a second review, once you have 2 approved reviews on this PR it should run

mirpedrol commented:
Regarding download, it's true that we have a bug which will be solved with the modifications on this PR. But the error message here is different. Let's first try to fix the stub run and see if it passes

LouisLeNezet reviewed (changes requested):
Just small fixes but otherwise good to go for me !
Thanks for addressing the different comments we had 😄 
I just have a question regarding how the stats are aggregated.
For now you don't pass them as a channel and I don't see how `STATS` output are obtained afterwards for example.

If you run `nf-core subworkflows lint` and `nf-core modules lint`, you might have a lot of warning to pay attention to. But that can be done in the future I guess.

nschan commented on subworkflows/local/mapping/map_sr/main.nf:
```diff
@@ -0,0 +1,46 @@
+        .map { meta, _reads, fasta -> [ meta, fasta ] }
+        .set { ch_fasta }
+
+    BAM_STATS(aln_to_assembly_bam_bai, ch_fasta)
```
They do not go into the report, but I think that having the bam stats in the results is useful, since they can be informative.

LouisLeNezet replied:
Okay !
Is it documented therefore in the output.md ?

nschan replied:
Yes https://github.com/nf-core/genomeassembler/blob/dev/docs/output.md#alignments lists them in output files.

nschan commented on tower.yml:
Execution on tower I assumed, I do not recall creating it and assumed it came with the template. I can remove it if it is not required.

nschan replied:
This seems to be part of the template: https://github.com/nf-core/tools/blob/main/nf_core/pipeline-template/tower.yml

nschan commented on workflows/genomeassembler.nf:
```diff
@@ -1,123 +1,240 @@
+        .collect()
+        .set { merqury_files }
+
+    Channel
```
This is collecting an output channel into a list of  `merqury` output files to pass into report. Those paths do not point to the copy in `publishDir`, but to the file in the process work directory. Maybe this is a confusion about the other k-mer `STATS` from `jellyfish`?

LouisLeNezet replied:
Yeah, I would need to try the pipeline to understand what's going on 🤔

nschan commented on nextflow.config:
```diff
@@ -9,126 +9,236 @@
+        shifter.enabled         = false
+        charliecloud.enabled    = false
+    }
+    wave {
```
Thanks, I will check for duplicates again, I think I messed up a merge at some point.

nschan commented on subworkflows/local/hifi/main.nf:
I do not really know what I was thinking when I set up those empty channels, probably that at some point I would add something that fills them, I think those can be removed. I would kind of like to keep this workflow to have ONT and HIFI consistent in terms of subworkflow organisation, but I think the two empty channels should go.

nschan commented on workflows/genomeassembler.nf:
```diff
@@ -1,123 +1,240 @@
-    CUSTOM_DUMPSOFTWAREVERSIONS (
-        ch_versions.unique().collectFile(name: 'collated_versions.yml')
-    )
+    REPORT(report_files, report_functions, nanoq_files, genomescope_files, quast_files, busco_files, merqury_files, Channel.fromPath("${params.outdir}/pipeline_info/nf_core_pipeline_software_versions.yml"))
```
It is very convenient for me to have the software versions in a yml for parsing into the report. I do see though that the way I do this right now is maybe not the best way. How would I go about creating the versions.yml in the actual module? Could I just  import `softwareVersionsToYAML` into the module and create the versions file in the work dir?

LouisLeNezet replied:
That's a good question.
Maybe you can create an issue on that and work on it for the next release 😉

nschan commented:
Thanks a lot for your review!

>  just have a question regarding how the stats are aggregated. For now you don't pass them as a channel and I don't see how STATS output are obtained afterwards for example.

I have added an explanation for this on the jellyfish subworkflow comment, let me know if that makes sense.

nschan commented on subworkflows/local/jellyfish/main.nf:
```diff
@@ -0,0 +1,57 @@
+
+    ch_versions = ch_versions.mix(GENOMESCOPE.out.versions)
+
+    STATS(kmers)
```
To illustrate, the output from jellyfish `STATS` is a file that looks like this:

```
Unique:    1855432275
Distinct:  2209215535
Total:     16107536483
Max_count: 4272501
```

nschan commented:
If there are no more open points please approve the PR @LouisLeNezet, so the full test runs. 
Here are things that I currently am aware of that should happen once this is approved:

- full test (AWS): make this pass in case it doesn't
- I will do a final small PR to add the date to `CHANGELOG.md`, before merging into master
- Do I create a release on github, or is that automated?
- Once I receive the DOI (after release) I will add that to `README.md`. 

is there anything that I missed?

Thanks again for taking the time to review this

LouisLeNezet reviewed (approved):
Thanks for adressing my last comments !
LGTM

LouisLeNezet commented:
You can look at the following checklist to see if you missed something
https://nf-co.re/docs/checklists/pipeline_release

nschan commented:
The AWS test does not start because of:
`/home/runner/work/_temp/6d46cf19-8e20-4786-aad2-fc6540a81f03.sh: line 106: done: command not found`
I do not really know what is wrong there, do you have an idea how I can fix this @mirpedrol ?

mirpedrol commented:
I triggered the workflow manually https://github.com/nf-core/genomeassembler/actions/runs/13584975634/job/37977891554

mirpedrol commented:
The test submits the pipeline to seqera platform, [this](https://cloud.seqera.io/orgs/nf-core/workspaces/AWSmegatests/watch/5T7sKptQcA3Qa6) is your run. If you don't have access to the nf-core space, let me know your user and I can add you

muffato commented:
👏🏼

LouisLeNezet commented:
Congrats on your first release !!

nschan commented:
Thanks a lot to all of you for the constructive reviews!
